li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaInstrumentedConsumerImpl.java

```diff
@@ -0,0 +1,583 @@
+import org.apache.kafka.common.TopicPartition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class LiKafkaInstrumentedConsumerImpl<K, V> implements Consumer<K, V>, EventHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(LiKafkaInstrumentedConsumerImpl.class);
+
+  private final ReadWriteLock delegateLock = new ReentrantReadWriteLock();
+  private final Properties baseConfig;
+  private final Map<String, String> translatedBaseConfig;
+  private final Function<Map<String, String>, Consumer<K, V>> consumerFactory;
+  private final String mdsUrl;
+  private final CountDownLatch initialConnectionLatch = new CountDownLatch(1);
+  private final MetricsProxy metricsProxy = new MetricsProxy();
+  private volatile Long closedAt = null;
+  private volatile Map<String, String> configOverrides;
+  private volatile Consumer<K, V> delegate;
+  private SimpleClient mdsClient;
+
+  private volatile Collection<String> subscribedTopics = null;
+  private volatile ConsumerRebalanceListener rebalanceListener = null;
+  private volatile Collection<TopicPartition> assignedPartitions = null;
+  private volatile Pattern subscriptionPattern = null;
+
+  public LiKafkaInstrumentedConsumerImpl(
+      Properties baseConfig,
+      Function<Map<String, String>, Consumer<K, V>> consumerFactory,
+      String mdsUrl
+  ) {
+    List<String> conversionIssues = new ArrayList<>(1);
+    this.baseConfig = baseConfig;
+    this.translatedBaseConfig = LiKafkaClientsUtils.toStringMap(baseConfig, conversionIssues);
+    this.consumerFactory = consumerFactory;
+    this.mdsUrl = mdsUrl;
+
+    if (!conversionIssues.isEmpty()) {
+      StringJoiner csv = new StringJoiner(", ");
+      conversionIssues.forEach(csv::add);
+      LOG.error("issues translating consumer config to strings: {}", csv);
+    }
+
+    mdsClient = new SimpleClient(
+        mdsUrl,
+        TimeUnit.MINUTES.toMillis(1),
+        TimeUnit.HOURS.toMillis(1),
+        translatedBaseConfig,
+        this
+    );
+
+    boolean tryFallback;
+    Exception issue = null;
+    try {
+      tryFallback = initialConnectionLatch.await(30, TimeUnit.SECONDS);
+    } catch (Exception e) {
+      tryFallback = true;
+      issue = e;
+    }
+
+    if (tryFallback) {
+      boolean delegateChanged = recreateDelegate(true);
+      if (delegateChanged) {
+        if (issue != null) {
+          LOG.error("exception waiting to contact {}, using user-provided configs as fallback", mdsUrl, issue);
+        } else {
+          LOG.error("unable to contact {} within timeout, using user-provided configs as fallback", mdsUrl);
+        }
+      } else if (issue != null) {
+        //we got interrupted waiting, but apparently connection to mds was successful?
+        LOG.warn("exception waiting on MDS connection, yet connection succeeded?", issue);
+      }
+    }
+  }
+
+  @Override
+  public void stateChanged(SimpleClient client, SimpleClientState oldState, SimpleClientState newState) {
```

Comment:
Does this actually handle config reload after client registration is done? For federated clients, each method was defined as synchronized for thread-safety. That's not needed here?

Reply:
called by a single event handler thread from the mds client. implementation is thread safe. config push added via a different method